### ISSUE TITLE
Aura authorship task MemoryAccessOutOfBounds while importing blocks

### ISSUE THREAD

**Issue by xlc** [paritytech/substrate#1363]
3 validators setup, a fork was created somehow (#1346 causing nodes banning each other), node 1 and 2 are in its own network and node 3 in a different one.

I wipe the node 3 volume and restart it hope it can catch up with with node 1 & 2 to solve the fork issue but getting this error while syncing

```
2019-01-08 03:05:36 Starting consensus session on top of parent 0x063bb8b542cd9cca7059d7649bb914e165da974702436d90f67ae1c7c04e2430
2019-01-08 03:06:38 Failed to construct block: Error(Inner(Error(Execution(Error(Wasmi(Trap(Trap { kind: MemoryAccessOutOfBounds })), State { next_error: None, backtrace: InternalBacktrace { backtrace: None } })), State { next_error: None, backtrace: InternalBacktrace { backtrace: None } })))
2019-01-08 03:06:38 Aura authorship task terminated unexpectedly. Restarting
```

Full logs

```
2019-01-08 03:02:55 Initialising Genesis block/state (state: 0xbc75…5026, header-hash: 0xdd02…ba81)
2019-01-08 03:02:56 Loaded block-time = 6 seconds from genesis on first-launch
2019-01-08 03:02:56 Loading GRANDPA authorities from genesis on what appears to be first startup.
2019-01-08 03:02:56 Best block: #0
2019-01-08 03:02:56 Local node address is: /ip4/0.0.0.0/tcp/30333/p2p/QmYcHeEWuqtr6Gb5EbK7zEhnaCm5p6vA2kWcVjFKbhApaC
2019-01-08 03:02:56 Using authority key 5G6tDRK9KSWLx3zL4gMnzh2K6QUNAb6EGCiyPMTAzCmt45A4
2019-01-08 03:02:56 Running Grandpa session as Authority 5G6tDRK9KSWLx3zL4gMnzh2K6QUNAb6EGCiyPMTAzCmt45A4
2019-01-08 03:02:56 Idle (0 peers), best: #0 (0xdd02…ba81)
2019-01-08 03:02:58 New external node address: /ip4/10.2.45.0/tcp/30333/p2p/QmYcHeEWuqtr6Gb5EbK7zEhnaCm5p6vA2kWcVjFKbhApaC
2019-01-08 03:02:58 New external node address: /ip4/10.2.46.0/tcp/30333/p2p/QmYcHeEWuqtr6Gb5EbK7zEhnaCm5p6vA2kWcVjFKbhApaC
2019-01-08 03:03:01 Syncing  7.2 bps, target=#2581 (2 peers), best: #36 (0x121f…6b28)
2019-01-08 03:03:06 Syncing  6.8 bps, target=#2581 (2 peers), best: #70 (0x2c4e…a265)
2019-01-08 03:03:11 Syncing  6.4 bps, target=#2582 (5 peers), best: #102 (0x943f…7ff4)
2019-01-08 03:03:16 Syncing  6.2 bps, target=#2583 (5 peers), best: #133 (0xd92d…68e1)
2019-01-08 03:03:21 Syncing  7.8 bps, target=#2584 (6 peers), best: #172 (0x3ea8…1f1a)
2019-01-08 03:03:26 Syncing  7.4 bps, target=#2585 (8 peers), best: #209 (0xc9f2…a686)
2019-01-08 03:03:31 Syncing  5.2 bps, target=#2586 (8 peers), best: #235 (0x215d…df85)
2019-01-08 03:03:36 Syncing  5.8 bps, target=#2586 (8 peers), best: #264 (0x1ee0…6e35)
2019-01-08 03:03:41 Syncing  6.2 bps, target=#2587 (8 peers), best: #295 (0x43c3…5ed2)
2019-01-08 03:03:46 Syncing  5.4 bps, target=#2588 (8 peers), best: #322 (0x0988…ac9a)
2019-01-08 03:03:51 Syncing  6.0 bps, target=#2589 (8 peers), best: #352 (0x30ab…b88e)
2019-01-08 03:03:56 Syncing  6.0 bps, target=#2590 (8 peers), best: #382 (0xea89…d727)
2019-01-08 03:04:01 Syncing  4.8 bps, target=#2591 (8 peers), best: #406 (0x3fc2…6b30)
2019-01-08 03:04:06 Syncing  4.8 bps, target=#2591 (8 peers), best: #430 (0x3092…f749)
2019-01-08 03:04:11 Syncing  5.6 bps, target=#2592 (8 peers), best: #458 (0x1179…d62e)
2019-01-08 03:04:16 Syncing  5.6 bps, target=#2593 (8 peers), best: #486 (0x94ef…c09a)
2019-01-08 03:04:21 Syncing  5.6 bps, target=#2594 (8 peers), best: #514 (0x223e…d67b)
2019-01-08 03:04:26 Syncing  6.0 bps, target=#2595 (8 peers), best: #544 (0x21ac…821e)
2019-01-08 03:04:31 Syncing  4.2 bps, target=#2596 (8 peers), best: #565 (0x6c80…e87a)
2019-01-08 03:04:36 Syncing  4.8 bps, target=#2596 (8 peers), best: #589 (0x361c…ef06)
2019-01-08 03:04:41 Syncing  5.0 bps, target=#2597 (8 peers), best: #614 (0x9160…2e47)
2019-01-08 03:04:46 Syncing  4.8 bps, target=#2598 (8 peers), best: #638 (0xd451…ec98)
2019-01-08 03:04:51 Syncing  4.4 bps, target=#2599 (8 peers), best: #660 (0x7d75…fb19)
2019-01-08 03:04:56 Syncing  5.8 bps, target=#2600 (8 peers), best: #689 (0x8808…d542)
2019-01-08 03:05:01 Syncing  4.8 bps, target=#2601 (8 peers), best: #713 (0x8555…d958)
2019-01-08 03:05:06 Syncing  3.6 bps, target=#2601 (8 peers), best: #731 (0x3258…04f2)
2019-01-08 03:05:11 Syncing  5.2 bps, target=#2602 (8 peers), best: #757 (0xf026…307c)
2019-01-08 03:05:16 Syncing  5.4 bps, target=#2603 (8 peers), best: #784 (0xc046…561e)
2019-01-08 03:05:21 Idle (8 peers), best: #812 (0xf61a…de92)
2019-01-08 03:05:26 Idle (8 peers), best: #837 (0xd620…701b)
2019-01-08 03:05:31 Idle (8 peers), best: #865 (0x3f70…e829)
2019-01-08 03:05:36 Starting consensus session on top of parent 0x063bb8b542cd9cca7059d7649bb914e165da974702436d90f67ae1c7c04e2430
2019-01-08 03:06:38 Failed to construct block: Error(Inner(Error(Execution(Error(Wasmi(Trap(Trap { kind: MemoryAccessOutOfBounds })), State { next_error: None, backtrace: InternalBacktrace { backtrace: None } })), State { next_error: None, backtrace: InternalBacktrace { backtrace: None } })))
2019-01-08 03:06:38 Aura authorship task terminated unexpectedly. Restarting
2019-01-08 03:06:38 Idle (8 peers), best: #896 (0x84eb…58fe)
2019-01-08 03:06:38 Idle (8 peers), best: #896 (0x84eb…58fe)
2019-01-08 03:06:38 Idle (8 peers), best: #896 (0x84eb…58fe)
2019-01-08 03:06:38 Idle (8 peers), best: #896 (0x84eb…58fe)
2019-01-08 03:06:38 Idle (8 peers), best: #896 (0x84eb…58fe)
2019-01-08 03:06:38 Idle (8 peers), best: #896 (0x84eb…58fe)
2019-01-08 03:06:38 Idle (8 peers), best: #896 (0x84eb…58fe)
2019-01-08 03:06:38 Idle (8 peers), best: #896 (0x84eb…58fe)
2019-01-08 03:06:38 Idle (8 peers), best: #896 (0x84eb…58fe)
2019-01-08 03:06:38 Idle (8 peers), best: #896 (0x84eb…58fe)
2019-01-08 03:06:38 Idle (8 peers), best: #896 (0x84eb…58fe)
2019-01-08 03:06:38 Idle (8 peers), best: #896 (0x84eb…58fe)
2019-01-08 03:06:38 Idle (8 peers), best: #896 (0x84eb…58fe)
2019-01-08 03:06:41 Idle (8 peers), best: #913 (0xcc6e…21ba)
2019-01-08 03:06:46 Idle (8 peers), best: #933 (0xe25c…f7a6)
2019-01-08 03:06:48 Starting consensus session on top of parent 0xf8376f4d6e8142743538d6df50a928474c8bea19c48bb16acdf10f6c7ab0489a
2019-01-08 03:07:48 Failed to construct block: Error(Inner(Error(Execution(Error(Wasmi(Trap(Trap { kind: MemoryAccessOutOfBounds })), State { next_error: None, backtrace: InternalBacktrace { backtrace: None } })), State { next_error: None, backtrace: InternalBacktrace { backtrace: None } })))
2019-01-08 03:07:48 Aura authorship task terminated unexpectedly. Restarting
2019-01-08 03:07:48 Idle (8 peers), best: #943 (0x842a…7e35)
2019-01-08 03:07:48 Idle (8 peers), best: #943 (0x842a…7e35)
2019-01-08 03:07:48 Idle (8 peers), best: #943 (0x842a…7e35)
2019-01-08 03:07:48 Idle (8 peers), best: #943 (0x842a…7e35)
2019-01-08 03:07:48 Idle (8 peers), best: #943 (0x842a…7e35)
2019-01-08 03:07:48 Idle (8 peers), best: #943 (0x842a…7e35)
2019-01-08 03:07:48 Idle (8 peers), best: #943 (0x842a…7e35)
2019-01-08 03:07:48 Idle (8 peers), best: #943 (0x842a…7e35)
2019-01-08 03:07:48 Idle (8 peers), best: #943 (0x842a…7e35)
2019-01-08 03:07:48 Idle (8 peers), best: #943 (0x842a…7e35)
2019-01-08 03:07:48 Idle (8 peers), best: #943 (0x842a…7e35)
2019-01-08 03:07:48 Idle (8 peers), best: #943 (0x842a…7e35)
2019-01-08 03:07:51 Idle (8 peers), best: #960 (0xe94d…a356)
2019-01-08 03:07:56 Idle (8 peers), best: #986 (0xe9a8…b9b7)
2019-01-08 03:08:00 Starting consensus session on top of parent 0xa7e33f6d6eecc4c8292d4fa04655f060ef4380bb3ca041ba1abe06ff7054e67a
2019-01-08 03:09:02 Failed to construct block: Error(Inner(Error(Execution(Error(Wasmi(Trap(Trap { kind: MemoryAccessOutOfBounds })), State { next_error: None, backtrace: InternalBacktrace { backtrace: None } })), State { next_error: None, backtrace: InternalBacktrace { backtrace: None } })))
2019-01-08 03:09:02 Aura authorship task terminated unexpectedly. Restarting
2019-01-08 03:09:02 Idle (8 peers), best: #1009 (0x49f8…afbd)
2019-01-08 03:09:02 Idle (8 peers), best: #1009 (0x49f8…afbd)
2019-01-08 03:09:02 Idle (8 peers), best: #1009 (0x49f8…afbd)
2019-01-08 03:09:02 Idle (8 peers), best: #1009 (0x49f8…afbd)
2019-01-08 03:09:02 Idle (8 peers), best: #1009 (0x49f8…afbd)
2019-01-08 03:09:02 Idle (8 peers), best: #1009 (0x49f8…afbd)
2019-01-08 03:09:02 Idle (8 peers), best: #1009 (0x49f8…afbd)
2019-01-08 03:09:02 Idle (8 peers), best: #1009 (0x49f8…afbd)
2019-01-08 03:09:02 Idle (8 peers), best: #1009 (0x49f8…afbd)
2019-01-08 03:09:02 Idle (8 peers), best: #1009 (0x49f8…afbd)
2019-01-08 03:09:02 Idle (8 peers), best: #1009 (0x49f8…afbd)
2019-01-08 03:09:02 Idle (8 peers), best: #1009 (0x49f8…afbd)
2019-01-08 03:09:02 Idle (8 peers), best: #1009 (0x49f8…afbd)
2019-01-08 03:09:06 Idle (8 peers), best: #1024 (0xdd88…6b86)
2019-01-08 03:09:11 Idle (8 peers), best: #1048 (0x875d…8ccb)
2019-01-08 03:09:12 Starting consensus session on top of parent 0x11ba3ea5aa38187ce89ce77f100dff05a8860e5e54521071584477023acf565d
```

**Comment by gnunicorn** [comment]
Aura Authorship fails because block execution fails, in this case `MemoryAccessOutOfBounds` in there implies that there is a bug in the runtime itself. I think it is correct that it is failing here, I don't see how the Authorship process is supposed to be able to recover if _block construction_ fails.

We could offer a recovery strategy inside BlockImporter that, when construction fails, attempts to construct again with fewer transactions (either attempting to identify the breaking one and drop that, or just pick a subset). Problem is that, in case of Aura, this quickly means we will run out of the time to actually build the block. So maybe we want to check the time left again and send an propose an empty block instead (to not be punished for being offline)?

**Comment by andresilva** [comment]
Can you confirm that nodes 1 and 2 are on blocks > #2500? Node 3 was initially targeting blocks > #2500 and for some reason it then assumed it was synced around block #800, this made the authorship task start trying to produce blocks (even though the node isn't synced yet). Not sure why this would case the memory access out of bounds in wasmi though.

**Comment by xlc** [comment]
Yes nodes 1 and 2 are on blocks > #2500.
It seems like the node 3 is still syncing (badly) when the error occurs that block number still increasing and much faster than 6s block time.

**Comment by rphmeier** [comment]
The aura authorship task doesn't author blocks unless the sync service says it's done syncing.

Of course there is never any definitive way to say if you are done syncing, so it has to be a heuristic only. Our heuristic may not be ideal but hard to say.

**Comment by gavofyork** [comment]
`MemoryAccessOutOfBounds` might just be the too-many-events-deposited issue we already saw and had to work around? @xlc is it possible that a lot of time (a few hours) has passed between consecutive blocks?

**Comment by andresilva** [comment]
@rphmeier But there's definitely something wrong going on here. The sync code uses as an heuristic the best seen block which was #2603 and then we switched to idle state after around block #800, a way for this to happen is if we lost the peer that was providing the best seen block but that doesn't seem to be the case here (and I guess there would be other peers at that height).

**Comment by rphmeier** [comment]
> `MemoryAccessOutOfBounds` might just be the too-many-events-deposited issue we already saw and had to work around? @xlc is it possible that a lot of time (a few hours) has passed between consecutive blocks?

@gavofyork if it tried to author at block # 800 when the actual height was closer to 2500, then there is quite a lot of time between the authoring time and the parent step.

**Comment by gavofyork** [comment]
@rphmeier yeah - that would probably do it.

@xlc if this is correct, then the issue should be fixed by #1371

**Comment by xlc** [comment]
I will try latest version and reopen this issue if it happened again.
